R/databricks.R

```diff
@@ -33,105 +33,191 @@
+           httpPath,
            workspace = Sys.getenv("DATABRICKS_HOST"),
            useNativeQuery = TRUE,
            driver = NULL,
            ...) {
     args <- databricks_args(
-      HTTPPath = HTTPPath,
+      httpPath = httpPath,
       workspace = workspace,
       useNativeQuery = useNativeQuery,
-      driver = driver
+      driver = driver,
+      ...
     )
-    args <- c(args, ...)
     inject(dbConnect(odbc(), !!!args))
   }
 )
 
-databricks_args <- function(HTTPPath,
+databricks_args <- function(httpPath,
                             workspace = Sys.getenv("DATABRICKS_HOST"),
                             useNativeQuery = FALSE,
-                            driver = NULL) {
-  if (nchar(workspace) == 0) {
-    stop("No Databricks workspace URL provided")
+                            driver = NULL,
+                            ...) {
+  host <- databricks_host(workspace)
+
+  args <- databricks_default_args(
+    driver = driver,
+    host = host,
+    httpPath = httpPath,
+    useNativeQuery = useNativeQuery
+  )
+  auth <- databricks_auth_args(host)
+  all <- c(args, auth, ...)
+
+  arg_names <- tolower(names(all))
+  if (!"authmech" %in% arg_names && !all(c("uid", "pwd") %in% arg_names)) {
+    warn(
+      c(
+        "x" = "Failed to detect ambient Databricks credentials.",
+        "i" = "Supply `uid` or `pwd` to authenticate manually."
+      ),
+      call = quote(DBI::dbConnect())
+    )
   }
-  hostname <- gsub("https://", "", workspace)
-  driver <- driver %||% default_databricks_driver()
 
-  # Check some standard Databricks environment variables. This is used to
-  # implement a subset of the "Databricks client unified authentication" model.
-  token <- Sys.getenv("DATABRICKS_TOKEN")
-  client_id <- Sys.getenv("DATABRICKS_CLIENT_ID")
-  client_secret <- Sys.getenv("DATABRICKS_CLIENT_SECRET")
-  cli_path <- Sys.getenv("DATABRICKS_CLI_PATH", "databricks")
+  all
+}
+
+databricks_default_args <- function(driver, host, httpPath, useNativeQuery) {
+  list(
+    driver = driver %||% databricks_default_driver(),
+    host = host,
+    httpPath = httpPath,
+    thriftTransport = 2,
+    userAgentEntry = databricks_user_agent(),
+    useNativeQuery = as.integer(useNativeQuery),
+    # Connections to Databricks are always over HTTPS.
+    port = 443,
+    protocol = "https",
+    ssl = 1
+  )
+}
+
+
 
-  user_agent <- paste0("r-odbc/", utils::packageVersion("odbc"))
+# Returns a sensible driver name even if odbc.ini and odbcinst.ini do not
+# contain an entry for the Databricks ODBC driver. For Linux and macOS we
+# default to known shared library paths used by the official installers.
+# On Windows we use the official driver name.
+databricks_default_driver <- function() {
+  default_paths <- databricks_default_driver_paths()
+  if (length(default_paths) > 0) {
+    return(default_paths[1])
+  }
+
+  fallback <- "Simba Spark ODBC Driver"
+  if (fallback %in% odbcListDrivers()$name) {
+    return(fallback)
+  }
```

Comment:
```suggestion
  fallbacks <- c("Databricks", "Simba Spark ODBC Driver")
  fallbacks <- intersect(fallbacks, odbcListDrivers()$name)
  if (len(fallbacks) > 1) {
    return(fallbacks[1])
  }
```

If we want to cover the majority of cases, I would add `Databricks` to a list of fallbacks to look for:

* `Databricks`
    * Driver name from the sample `odbcinst.ini` file shipped with the Posit drivers (for Linux/macOS)
    * Driver name as written by Desktop Pro on [Linux/macOS](https://github.com/rstudio/rstudio/blob/d9c0b090d49752fe60e7a2ea4be3123cc3feeb6c/src/cpp/session/modules/SessionConnectionsInstaller.R#L436-L441) and [Windows](https://github.com/rstudio/rstudio/blob/d9c0b090d49752fe60e7a2ea4be3123cc3feeb6c/src/cpp/session/modules/SessionConnectionsInstaller.R#L482-L486)
* `Simba Spark ODBC Driver` is the name of the driver on Windows when installed from the [Databricks download](https://www.databricks.com/spark/odbc-drivers-download)

The driver name is dependent on how the user has configured the system so it may not be the Posit driver.